rfcs/2025-shadow-dom-support.md

```diff
@@ -0,0 +1,128 @@
+<!-- Copyright 2025 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License. -->
+
+- Start Date: 2025/1/28
+- RFC PR: exploration PRs: https://github.com/adobe/react-spectrum/pull/6046
+- Authors: Rob Snow
+
+# Improving React Aria Shadow DOM Support
+
+## Summary
+
+This RFC outlines a plan for progressive enhancement of Shadow DOM support in React Aria, React Aria Components, and React Spectrum S2.
+Shadow DOM support can mean many things; open root vs closed root, single containing shadow root, or multitude of individual components.
+We can improve across all of these.
+
+
+## Motivation
+
+As Web Components have become more prominent and more libraries and applications use them, we have encountered friction for people trying to adopt or use our libraries alongside others.
+
+Some examples of these are:
+- [Dialog's focus management and work with 3rd party dialogs](https://github.com/adobe/react-spectrum/issues/5314)
+- [Video Controls are not respected when using FocusScope](https://github.com/adobe/react-spectrum/issues/6729)
+- [FocusScope not working when used inside shadowRoot](https://github.com/adobe/react-spectrum/issues/1472)
+- [ariaHideOutside incorrect behavior inside shadow DOM.](https://github.com/adobe/react-spectrum/issues/6133)
+- [usePress is not work in shadowRoot](https://github.com/adobe/react-spectrum/issues/2040)
+- [useOverlay Click Outside in Shadow-DOM context](https://github.com/adobe/react-spectrum/issues/3970)
+
+We have also had a contribution to solve some of the issues. While this is useful, we would like it to feel less hacky and more importantly, we'd like to incorporate the support into our daily lives in as easy as way as possible.
+
+
+## Detailed Design
+
+As mentioned earlier, there are proposed parts to this initiative:
+
+1. Custom React Testing Library Renderer
```

Comment:
We do not write special tests for shadow DOM. All our component testing is done using playwright's component testing. We run the same tests both with and without shadow DOM.

Reply:
I'm not sure what this comment has to do with us? The RFC is for what we need to do to support shadow dom in our library.

I'm also suggesting that we run the same tests both with and without shadow DOM
> we should create a custom React Testing Library renderer for our unit tests which can wrap each test's rendered dom in a shadow root.
> ... without needing to write many specific tests.

There will need to be specific tests as well, but they can be added as edge cases are discovered. For now I'd like to get all of our current tests running both ways.

Reply:
> I'm not sure what this comment has to do with us?

Not much. This was just me sharing how we handle this situation.